# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1221978112 -7200
#      Sun Sep 21 08:21:52 2008 +0200
# Branch release-3-0-x
# Node ID 74802398931b8cf2dc664eaaa244de78988442bf
# Parent  79d5312ab8e6bda4fec3720e6c9c361bf5c1a1e9
fix blkdiag to not rely on Matlab-incompatible behaviour

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,12 @@
+2008-09-18  Jaroslav Hajek <highegg@gmail.com>
+
+	* general/blkdiag.m: Skip assignment if rhs is empty.
+
 2008-09-18  Kris Thielemans  <kris.thielemans@csc.mrc.ac.uk>
 
 	* image/imshow.m: Fix for display_range.
 	
 	* general/interpn.m: Fixe for extrapval and documentation of 
 	extrapval.
 
 2008-09-19  David Bateman  <dbateman@free.fr>
diff --git a/scripts/general/blkdiag.m b/scripts/general/blkdiag.m
--- a/scripts/general/blkdiag.m
+++ b/scripts/general/blkdiag.m
@@ -44,17 +44,20 @@ function retval = blkdiag (varargin)
 
   ## size is an option for cellfun, but it's a bit different from
   ## calling size directly.
   tmp = cell2mat (cellfun (@size, varargin', "UniformOutput", false));
   csz = cumsum ([0 0; tmp], 1);
   retval = zeros (csz(end,:));
 
   for p = 1:nargin
-    retval((csz(p,1)+1):csz(p+1,1),(csz(p,2)+1):csz(p+1,2)) = varargin{p};
+    vp = varargin{p};
+    if (! isempty (vp))
+      retval((csz(p,1)+1):csz(p+1,1),(csz(p,2)+1):csz(p+1,2)) = vp;
+    endif
   endfor
 
 endfunction
 
 # regular tests
 %!assert(blkdiag(1,ones(2),1),[1,0,0,0;0,1,1,0;0,1,1,0;0,0,0,1])
 %!assert(blkdiag([1,2],[3,4],[5,6]),[1,2,0,0,0,0;0,0,3,4,0,0;0,0,0,0,5,6])
 %!assert(blkdiag([1,2],[3;4],[5,6]),[1,2,0,0,0;0,0,3,0,0;0,0,4,0,0;0,0,0,5,6])
